_posts/2019-02-25-practice-oop-to-front-end-universal-state-module.md

```diff
@@ -0,0 +1,291 @@
+- Optional Event System
+- Support Stateless Model
+- Support React/Vue/Angular
+
+### Decorators
+
+`usm` provides decorator `@state` to wrap a variable with a state, and decorator `@action` is used to wrap a function that changes state (the last parameter passed in by the function is always the current state object).
+
+```js
+class Shop extends Module {
+  @state goods = [];
+  @state status = 'close';
+
+  @action
+  operate(item, status, state) {
+    state.goods.push(item);
+    state.status = status;
+  }
+  // call function -> this.operate({ name: 'fruits', amount: 10 }, 'open');
+}
+```
+
+### Module lifecycle
+
+`usm` provides these lifecycle events:
+
+- `moduleWillInitialize`
+- `moduleWillInitializeSuccess`
+- `moduleDidInitialize`
+- `moduleWillReset`
+- `moduleDidReset`
+
+The order in which they are run is shown in the following flow chart:
+
+![lifecycle](/integration-blog/assets/2019-02-25-practice-oop-to-front-end-universal-state-module/usm_lifecycle.png)
+
+These module lifecycles can be used to coordinate module initialization dependencies.
+
+```js
+class TodoList extends Module {  
+  async moduleWillInitialize() {
+    console.log(
+      'TodoList -> moduleWillInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+
+  async moduleWillInitializeSuccess() {
+    console.log(
+      'TodoList -> moduleWillInitializeSuccess',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+      );
+  }
+
+  async moduleDidInitialize() {
+    console.log(
+      'TodoList -> moduleDidInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+      );
+  }
+}
+
+
+class App extends Module{
+  async moduleWillInitialize() {
+    console.log(
+      'App -> moduleWillInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+
+  async moduleWillInitializeSuccess() {
+    console.log(
+      'App -> moduleWillInitializeSuccess',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+
+  async moduleDidInitialize() {
+    console.log(
+      'App -> moduleDidInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+}
+
+const todoList = new TodoList();
+
+const app = App.create({
+  modules: {
+    todoList
+  }
+});
+```
+console.log reulst:
+
+```
+App -> moduleWillInitialize pendding: false ready: false
+TodoList -> moduleWillInitialize pendding: false ready: false
+TodoList -> moduleWillInitializeSuccess pendding: true ready: false
+App -> moduleWillInitializeSuccess pendding: true ready: false
+TodoList -> moduleDidInitialize pendding: false ready: true
+App -> moduleDidInitialize pendding: false ready: true
+```
+
+### An ideal architecture
+
+![flow chart](/integration-blog/assets/2019-02-25-practice-oop-to-front-end-universal-state-module/flow_chart.png)
+
+In a complex front-end application, this may be a more typical modular architecture design that contains the following sections:
```

Comment:
In a complex front-end application, a typical modular architecture may contain the following: